Sources/SwiftRefactor/ExpandEditorPlaceholder.swift

```diff
@@ -0,0 +1,339 @@
+/// type-string. If this ends up not the case for some reason, we can consider
+/// adding escaping for '##'.
+///
+/// The type string provided in the placeholder (preferring
+/// `type-for-expansion-string`), is parsed into a syntax node. If that node is
+/// a `FunctionTypeSyntax` then the placeholder is expanded into a
+/// `ClosureExprSyntax`. Otherwise it is expanded as is, which is also the case
+/// for when only a display string is provided.
+///
+/// ## Function Typed Placeholder
+/// ### Before
+/// ```swift
+/// <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>
+/// ```
+///
+/// ### After
+/// ```swift
+/// { someInt in
+///   <#T##String#>
+/// }
+/// ```
+///
+/// ## Other Type Placeholder
+/// ### Before
+/// ```swift
+/// <#T##Int#>
+/// ```
+///
+/// ### After
+/// ```swift
+/// Int
+/// ```
+///
+/// ## No Type Placeholder
+/// ### Before
+/// ```swift
+/// <#anything here#>
+/// ```
+///
+/// ## After
+/// ```swift
+/// anything here
+/// ```
+public struct ExpandEditorPlaceholder: RefactoringProvider {
+  public static func refactor(syntax token: TokenSyntax, in context: Void) -> ExprSyntax? {
+    guard let placeholder = EditorPlaceholderData(token: token) else {
+      return nil
+    }
+
+    let expanded: ExprSyntax
+    switch placeholder {
+    case let .basic(text):
+      var parser = Parser(String(text))
+      expanded = ExprSyntax.parse(from: &parser)
+    case let .typed(text, type):
+      if let functionType = type.as(FunctionTypeSyntax.self) {
+        expanded = ExprSyntax(functionType.closureExpansion)
+      } else {
+        var parser = Parser(String(text))
+        expanded = ExprSyntax.parse(from: &parser)
+      }
+    }
+
+    return expanded
+      .with(\.leadingTrivia, token.leadingTrivia)
+      .with(\.trailingTrivia, token.trailingTrivia)
+  }
+}
+
+/// If a function-typed placeholder is the argument to a non-trailing closure
+/// call, expands it and any adjacent function-typed placeholders to trailing
+/// closures on that call. All other placeholders will expand as per
+/// `ExpandEditorPlaceholder`.
+///
+/// ## Before
+/// ```swift
+/// foo(closure1: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, normalArg: <#T##Int#>, closure2: { ... }, closure3: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, closure4: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>
+/// ```
+///
+/// ## `closure3` or `closure4` Expansion
+/// ```swift
+/// foo(closure1: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, normalArg: <#T##Int#>, closure2: { ... }) { someInt in
+///   <#T##String#>
+/// } closure2: { someInt in
+///   <#T##String#>
+/// }
+/// ```
+///
+/// Expansion on `closure1` and `normalArg` is the same as `ExpandEditorPlaceholder`.
+public struct ExpandEditorPlaceholders: RefactoringProvider {
+  public static func refactor(syntax token: TokenSyntax, in context: Void) -> ExprSyntax? {
```

Comment:
This is a bit of a weird API at the moment. We take a `TokenSyntax` but may return a `FunctionCallExprSyntax` if we convert the call itself to trailing closure syntax. Would love to hear opinions on this 😅

Reply:
Any reason not to have it be like `CallToTrailingClosures` and have it take the function call with a starting index?

Reply:
Just that it seemed a little weird to force clients to check whether they have a `TokenSyntax` within a `FunctionCallExprSyntax` themselves. Though having said that, I'm not sure there's any way at the moment for the client to know that the *parent* was replaced rather than the `TokenSyntax` they passed. So maybe it is better to go the `CallToTrailingClosures` way.

Reply:
My current plan here is to add a new protocol `EditRefactoringProvider`. It would have a single:
```
  static func refactor(syntax: Self.Input, in context: Self.Context) -> [Edit]
```

`RefactoringProvider` would itself be an `EditRefactoringProvider`, providing a default implementation of `refactor` that removes the input node and replaces it with the returned node.

Then `ExpandEditorPlaceholders` would no longer be a `RefactoringProvider`, but rather a `EditRefactoringProvider`. Possibly `ExpandEditorPlaceholder` too, haven't decided on that one yet. It really does just return text, so it probably shouldn't be a `RefactoringProvider`.